### DISCUSSION TITLE
Data Encapsulation

### DISCUSSION THREAD

**Discussion by asiFarran** [mikro-orm/mikro-orm#1817] [Q&A] [answered]
Hi,

First of all, kudos on this great project!

My team and I are evaluating mikro-orm for use in a existing largish enterprise project - swapping out sequelize.

We've created a POC and are very happy with the experience so far but have found friction around data encapsulation.

In our codebase, entity **data properties** are private and have matching accessors:

```
@Property()
private _name!: string;

get name() {
  return this._name;
}

/* optional, some props are readonly and can only be mutated through methods.
  ids for example are always readonly */

set name (val:string){ 
  this._name = val;
}
```

Using an appropriate  `NamingStrategy` and a bit of customization around _populate_ (so that we can use the public field names rather than the _underscored ones) it all works pretty well! 

However, where we're running into issues is when modeling bi-directional relationships. For e,g:

- We cannot correctly specify the `mappedBy` reference as the 'real' property is private and so not a `key of T`
- We've tried to see what happens if we modify a bit the decorators so that the property names line up, but then run into other issues deeper in the guts of metadata discovery , where we're less confident to play around right now without deeper understanding.

Does anyone have some insight to share on a way to tackle this encapsulation requirement with mikro-orm , either as a pattern or as pointers on the appropriate place in the code where support for this could be added most safely?

Thanks much

(p.s. note that we're currently using tsmorph but have no problem to swap to entity schema if that makes a difference).

**Comment by B4nan** [comment, marked as answer]
> We cannot correctly specify the mappedBy reference as the 'real' property is private and so not a key of T

Sounds like a TS issue, so easily workaroundable via type casting (or you can pass both `entity` and `mappedBy` as strings which won't be type checked).

**Reply by asiFarran** [reply]
Thanks for the quick reply @B4nan , we're very much trying to avoid using magic strings. 
Also, I'd like to end up with a solution that:
1. Does not the require the developer to use the _underscored private field names anywhere outside the entity itself
2. Is not relying on the developer using one specific style of declaring the binding (i.e `@OneToMany({ mappedBy: book => book.author })` and `@OneToMany({ entity: () => Book, mappedBy: 'author', orphanRemoval: true })` will both work)

**Reply by B4nan** [reply]
> Does not the require the developer to use the _underscored private field names anywhere outside the entity itself

Well but that can never work, as you need to setup relations where you want to use those `_` fields. 

> Is not relying on the developer using one specific style of declaring the binding 

Same problem, if you want that, don't use private modifier 🤷 The callback syntax will always require public properties, that is how TS works.

I am not saying we cannot do better, but those two seems to be impossible to workaround with current state of things (and are both pretty much TS problems, nothing in our code). Maybe you can propose how would you expect this to work (on the surfacing API level)? AFAIK you can't infer private properties from object in TS.

**Reply by asiFarran** [reply]
OK, we seem to have found a compromise in sub-classing some of the decorators and making the adjustments there , as well as using a custom base entity repository.

We're still flushing this out but when we have a clean example I'll post a link back to it here in case its of interest to anyone else.
